File: python/06.py
import re

filename = "06.txt"

regex = re.compile(r'^(turn on|toggle|turn off)\s+(\d+,\d+)\s.*\s(\d+,\d+)$')

lights = [[0 for x in range(1000)] for x in range(1000)]
instructions = [x.strip() for x in open(filename).readlines()]

def turn_on(lights, start, finish):
    (start_x, start_y) =map(int, start.split(","))
    (finish_x, finish_y) = map(int, finish.split(","))

    for x in range(start_x, finish_x + 1):
        for y in range(start_y, finish_y + 1):
            lights[x][y] = lights[x][y] + 1

def turn_off(lights, start, finish):
    (start_x, start_y) =map(int, start.split(","))
    (finish_x, finish_y) = map(int, finish.split(","))

    for x in range(start_x, finish_x + 1):
        for y in range(start_y, finish_y + 1):
            lights[x][y] = max(0, lights[x][y] - 1)

def toggle(lights, start, finish):
    (start_x, start_y) =map(int, start.split(","))
    (finish_x, finish_y) = map(int, finish.split(","))

    for x in range(start_x, finish_x + 1):
        for y in range(start_y, finish_y + 1):
            lights[x][y] = lights[x][y] + 2


for i in instructions:
    m = regex.match(i)
    if not m:
        break

    (instruction, start, finish) = m.groups()

    if instruction.strip() == "turn on":
        turn_on(lights, start, finish)
    elif instruction.strip() == "toggle":
        toggle(lights, start, finish)
    elif instruction.strip() == "turn off":
        turn_off(lights, start, finish)

lumosity = 0

for x in range(1000):
    for y in range(1000):
        lumosity = lumosity + lights[x][y]

print lumosity

File: javascript/06.js
var fs = require('fs');
var filename = '06.txt';

var grid = [];
var grid_size = 1000;

// pre-fill our grid system
for (var i = 0; i < grid_size; i++) {
  grid.push(Array.apply(null, Array(grid_size)).map(function () { return 0; }));
}

var directions_p1 = {
  "turn on" : turnOn,
  "turn off" : turnOff,
  "toggle" : toggle
} ;

var instructions = fs.readFileSync(filename).toString().trim().split('\n');
var regex = /^(turn on|toggle|turn off)\s+(\d+,\d+)\sthrough\s(\d+,\d+)$/ig;

function stringToCoords(s) {
  var parts = s.split(',');
  return { 'x': parseInt(parts[0]), 'y': parseInt(parts[1]) };
}

function turnOn(lights, start, finish) {
  for (var x = start.x; x <= finish.x; x++) {
    for (var y = start.y; y <= finish.y; y++) {
      lights[x][y] = 1;
    }
  }
}

function turnOff(lights, start, finish) {
  for (var x = start.x; x <= finish.x; x++) {
    for (var y = start.y; y <= finish.y; y++) {
      lights[x][y] = 0;
    }
  }
}

function toggle(lights, start, finish) {
  for (var x = start.x; x <= finish.x; x++) {
    for (var y = start.y; y <= finish.y; y++) {
      lights[x][y] = (lights[x][y] ? 0 : 1);
    }
  }
}

instructions.forEach(function (instruction) {
  var match = /^(turn on|toggle|turn off)\s+(\d+,\d+)\sthrough\s(\d+,\d+)$/ig.exec(instruction.trim());
  directions_p1[match[1]](grid, stringToCoords(match[2]), stringToCoords(match[3]));
})

console.log(grid.reduce(function (total, row) {
  return total + row.reduce(function (rowTotal, column) {
    return rowTotal + column;
  }, 0)
}, 0));
